### ISSUE TITLE
Adding search feature to Select Component

### ISSUE THREAD

**Issue by batuhantozun** [Trendyol/baklava#265]
[Figma](https://www.figma.com/file/RrcLH0mWpIUy4vwuTlDeKN/Baklava-Design-Guide?node-id=2562%3A21640)

We need to add searchable feature to our select component.
Details will be here...

**Comment by leventozen** [comment]
## Search Feature in Select Component Design Meeting Report

Date: 26.10.2022 15:30 (UTC+3)

* We don't need to a have a `Search component` for this feature.
* We'll need an enhancement on the input component to have a "search input" version (with a search icon at left and clear icon at right to clear search inputs.)
* Search input will be sticky.
* When select opened, search input will be focused.
* When I type, I shouldn't see selected items. Otherwise, those should be at top.

Design work will start according to those decisions then implementation will be planned.

**Comment by pyilmazer** [comment]
Implementation details question : Will there be any count limit to separate in memory items flow from server side processing flow? This behaviour may be determined with a prop. Ex: Apply in memory items flow when item count is less than 100.

**Comment by batuhantozun** [comment]
We need to decide some details regarding to our update meeting. Here is the details;

- Adding loading state to input component
- Adding no data status to search results
- Can we move search icon to right of input component?

@buseselvi FYI

**Comment by musaakkayaa** [comment]
## Design

[Figma](https://www.figma.com/file/RrcLH0mWpIUy4vwuTlDeKN/Baklava-Design-Guide?node-id=25%3A3606&t=2nNj1LncV2RG17sa-0)

## Implementation

General usage example:

```html 
<bl-select>
    <bl-select-option value="tr">Turkey</bl-select-option>
    <bl-select-option value="nl">Netherlands</bl-select-option>
</bl-select>
```

### Rules

## API Reference:

### `bl-select` component

Description

#### Attributes

| Attribute | Description | Default Value |
| --------------- | --------------- | --------------- |
| label (`string`) | Sets the label value (optional) | - |
| placeholder(`string`) | Sets the placeholder value. If left blank, the label value (if specified) is set as placeholder. (optional) | - |
| size(`string`) | Sets the size value. Select component's height value will be changed accordingly (large, medium, small) | medium |
| required (`boolean`) | When option is not selected, shows component in error state | false |
| disabled (`boolean`) | Shows the component in disabled state | false |
| multiple (`boolean`) | Allows multiple options to be selected | false |
| label-fixed (`boolean`) | Makes label as fixed positioned | false |
| help-text (`string`) | Adds help text(optional) | - |
| invalid-text (`string`) | Set custom error message(optional) | - |
| search-bar (`boolean`) | Show search text for options(optional) | - |
| search-bar-placeholder (`string`) | Search text placeholder(optional) | - |
| search-not-found-text (`string`) | When options not found, shown this text | Not Found |
| search-bar-loading-state (`boolean`) | Search bar loading state for fetch request | false |
| search-filterable (`boolean`) | Search in the option list | true |

#### Slots

| Name | Description | Default Content |
| --------------- | --------------- | --------------- |
| `default` slot | Content inside select  | - |
| `search-not-found` slot | Custom show search not found layout | - |

#### Events

| Event | Description |
| --------------- | --------------- |
| `bl-select` | Will be triggered when click option |
| `bl-search` | Will be triggered every time search bar key down |

**Comment by muratcorlu** [comment]
In this issue we can focus search feature only. So you can remove unchanged attributes/events from the ADR.

I would like to read some information about how search will work in some different use-cases. Can we add some code examples to the ADR?

* Will it work with a static of list in the dom?
* Will user able to do a search on server-side with this functionality? If so, how?
* We accept so much text attributes to customize messages. I think this will cause so much code repetition and crowd in case you just want to use select in a different language then English. Maybe we should consider this feature together with our [localization task](#398)

Let's have a conclusion here first then continue implementation. That will be easier, I think.

We can also arrange a meeting if it's needed.

**Comment by muratcorlu** [comment]
## Design Meeting Report

Date: 2023-04-18 14:00 GMT+3

---

We had another discussion about this while talking about "Select All" feature (#520) Here are the search related decisions:

* Search functionality will be optional in contrast to "Select All".
* Search field will be sticky inside the popover.
* Popover height will still include up to 6 items even if the search field is visible. That means popover height in search mode will be able to bigger than single select mode.

**Comment by buseselvi** [comment]
Here is the final [design documentation link](https://www.figma.com/file/RrcLH0mWpIUy4vwuTlDeKN/Baklava-Design-Guide?type=design&node-id=4378%3A23660&t=E8wULDVAUnoZCOwA-1) 🚀

**Comment by AykutSarac** [comment]
Good job! Can you share your recommended solution to swap between the current behaviour of displaying selected options and the transformation after clicking to search among options? (like what's rendered before and after etc.)

**Comment by erbilnas** [comment]
Yes, I agree with you on that matter. Regarding our localization considerations, I believe the 'search-not-found' slot addresses it. However, for the placeholder, I think we can't resolve that using a slot since it needs to be within the input.

**Comment by leventozen** [comment]
Let's get a comment from @buseselvi about it. I personally preferring not to use a slot for this manner to have generic displaying. However, I think we can give a little bit flexibility after we successfully apply our decisions on that matter.

**Comment by buseselvi** [comment]
Hello, we agreed not to use slots here 🚀  

Also, I did a little research about the clear selected items and clear search we discussed. I couldn't find an example with clear search (for multiple select). So, let's not add a clear icon button to delete the text written in search and leave the clear button as it is (removing selected items). 
We can keep showing search icon while typing like in the design, but we can still show "Clear Search" button when there is "No Data Found". 🙏🏻 @erbilnas
